crates/sui-indexer/src/indexer_reader.rs

```diff
@@ -1483,99 +1404,117 @@ impl IndexerReader {
         })
     }
 
-    pub async fn get_coin_metadata_in_blocking_task(
-        &self,
-        coin_struct: StructTag,
-    ) -> Result<Option<SuiCoinMetadata>, IndexerError> {
-        self.spawn_blocking(move |this| this.get_coin_metadata(coin_struct))
+    pub async fn get_singleton_object(&self, type_: &StructTag) -> Result<Option<Object>> {
+        use diesel_async::RunQueryDsl;
+
+        let mut connection = self.pool.get().await?;
+
+        let object = match objects::table
+            .filter(objects::object_type.eq(type_.to_canonical_string(/* with_prefix */ true)))
```

Comment:
`object_type` does not have its own index, this query will be a full-scan and very slow, shall we keep the logic of `get_single_obj_id_from_package_publish` & `package_obj_type_cache`?

Reply:
Graphql does have the ability to do a singleton lookup based on type. I was under the impression we did have sufficient indexes for this because this is essentially how graphql does this

Reply:
Note that GraphQL makes this query on `objects_history` which does have an index on just `type`:

https://github.com/MystenLabs/sui/blob/1ebf5f8a395ca00f7cdcfa19e988975d8ccc15a3/crates/sui-indexer/migrations/pg/2023-08-19-044023_objects/up.sql#L72

I think @gegaowp is right that today, this query will result in a full table scan. `objects` does have an index that you can leverage:

https://github.com/MystenLabs/sui/blob/1ebf5f8a395ca00f7cdcfa19e988975d8ccc15a3/crates/sui-indexer/migrations/pg/2023-08-19-044023_objects/up.sql#L40

But it needs additional filters -- I'll put up a PR for this. It was created this way so that the same index could be used to support filtering by just the type's package, module, name and then the full type.

Reply:
Fixed in #19247